// @allurereport/service/dist/legacyService.js
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m") throw new TypeError("Private method is not writable");
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _AllureLegacyServiceClient_client, _AllureLegacyServiceClient_url, _AllureLegacyServiceClient_reportUrl;
import { readFile } from "node:fs/promises";
import { join as joinPosix } from "node:path/posix";
import { createServiceHttpClient } from "./utils/http.js";
const ASSET_MAX_FILE_SIZE = 200 * 1024 * 1024;
const UPLOAD_CONTENT_TYPE = "application/octet-stream";
const createUploadBlob = (content) => new Blob([content], { type: UPLOAD_CONTENT_TYPE });
const createReportFileUrl = (reportUrl, reportFilename) => {
    const fileUrl = new URL(reportUrl);
    fileUrl.pathname = joinPosix(fileUrl.pathname, reportFilename);
    fileUrl.search = "";
    fileUrl.hash = "";
    return fileUrl.toString();
};
const createFallbackReportUrl = (baseUrl, reportUuid) => {
    const reportUrl = new URL(`${baseUrl}/`);
    reportUrl.pathname = joinPosix(reportUrl.pathname, reportUuid);
    return reportUrl;
};
export class AllureLegacyServiceClient {
    constructor(config) {
        this.config = config;
        _AllureLegacyServiceClient_client.set(this, void 0);
        _AllureLegacyServiceClient_url.set(this, void 0);
        _AllureLegacyServiceClient_reportUrl.set(this, void 0);
        if (!config?.accessToken) {
            throw new Error("Allure service access token is required");
        }
        const { url } = JSON.parse(atob(config.accessToken.split(".")[1]));
        __classPrivateFieldSet(this, _AllureLegacyServiceClient_url, url.replace(/\/$/, ""), "f");
        __classPrivateFieldSet(this, _AllureLegacyServiceClient_client, createServiceHttpClient(__classPrivateFieldGet(this, _AllureLegacyServiceClient_url, "f"), config.accessToken), "f");
    }
    async downloadHistory(payload) {
        const { branch, limit } = payload ?? {};
        const { history } = await __classPrivateFieldGet(this, _AllureLegacyServiceClient_client, "f").get("/projects/history", {
            params: {
                limit: limit ? encodeURIComponent(limit) : undefined,
                branch: branch ? encodeURIComponent(branch) : undefined,
            },
        });
        return history;
    }
    async createReport(payload) {
        const { reportName, reportUuid, branch } = payload;
        const { url } = await __classPrivateFieldGet(this, _AllureLegacyServiceClient_client, "f").post("/reports", {
            body: {
                reportName,
                reportUuid,
                branch,
            },
        });
        __classPrivateFieldSet(this, _AllureLegacyServiceClient_reportUrl, new URL(url), "f");
        return __classPrivateFieldGet(this, _AllureLegacyServiceClient_reportUrl, "f");
    }
    async completeReport(payload) {
        const { reportUuid, historyPoint } = payload;
        return __classPrivateFieldGet(this, _AllureLegacyServiceClient_client, "f").post(`/reports/${reportUuid}/complete`, {
            body: {
                historyPoint,
            },
        });
    }
    async deleteReport(payload) {
        const { reportUuid, pluginId = "" } = payload;
        return __classPrivateFieldGet(this, _AllureLegacyServiceClient_client, "f").post(`/reports/${reportUuid}/delete`, {
            body: {
                pluginId,
            },
        });
    }
    async addReportAsset(payload) {
        const { filename, file, filepath, signal } = payload;
        if (!file && !filepath) {
            throw new Error("File or filepath is required");
        }
        let content = file;
        if (!content) {
            content = signal ? await readFile(filepath, { signal }) : await readFile(filepath);
        }
        if (content.length > ASSET_MAX_FILE_SIZE) {
            throw new Error(`Asset size exceeds the maximum allowed size of ${ASSET_MAX_FILE_SIZE / (1024 * 1024)}MB`);
        }
        const form = new FormData();
        form.set("filename", filename);
        form.set("file", createUploadBlob(content), filename);
        return __classPrivateFieldGet(this, _AllureLegacyServiceClient_client, "f").post("/assets/upload", {
            body: form,
            headers: { "Content-Type": "multipart/form-data" },
            ...(signal ? { signal } : {}),
        });
    }
    async addReportFile(payload) {
        const { reportUuid, filename, file, filepath, pluginId, signal } = payload;
        const reportFilename = pluginId ? joinPosix(pluginId, filename) : filename;
        if (!file && !filepath) {
            throw new Error("File or filepath is required");
        }
        let content = file;
        if (!content) {
            content = signal ? await readFile(filepath, { signal }) : await readFile(filepath);
        }
        if (content.length > ASSET_MAX_FILE_SIZE) {
            throw new Error(`Report file size exceeds the maximum allowed size of ${ASSET_MAX_FILE_SIZE / (1024 * 1024)}MB`);
        }
        const form = new FormData();
        form.set("filename", reportFilename);
        form.set("file", createUploadBlob(content), reportFilename);
        await __classPrivateFieldGet(this, _AllureLegacyServiceClient_client, "f").post(`/reports/${reportUuid}/upload`, {
            body: form,
            headers: { "Content-Type": "multipart/form-data" },
            ...(signal ? { signal } : {}),
        });
        return createReportFileUrl(__classPrivateFieldGet(this, _AllureLegacyServiceClient_reportUrl, "f") ?? createFallbackReportUrl(__classPrivateFieldGet(this, _AllureLegacyServiceClient_url, "f"), reportUuid), reportFilename);
    }
}
_AllureLegacyServiceClient_client = new WeakMap(), _AllureLegacyServiceClient_url = new WeakMap(), _AllureLegacyServiceClient_reportUrl = new WeakMap();
